File: clojure/005/005.clj
(defn remove_first
  ([x, coll]
    (let [[n m] (split-with #(not= x %) coll)]
      (concat n (rest m))
    )
  )
)

(defn find_first
  ([pred, coll]
    (first (filter pred coll))
  )
)

(defn overlap
  ([a, b]
    (let [duplicates_removed
      (loop [x a y b]
        (if (empty? x)
          y
          (recur (rest x) (remove_first (first x) y))
        )
      )]
      (sort (concat a duplicates_removed))
    )
  )
)

(defn prime_factors
  ([x]
    (let [v (find_first #(= (mod x %) 0) (range 2 x))]
      (if (= v nil)
        (list x)
        (concat (list v) (prime_factors (/ x v)))
      )
    )
  )
)

(println (reduce *
  (let [prime_factor_lists
    (map #(prime_factors %) (range 1 21))]
      (loop [remaining_prime_factor_lists prime_factor_lists factors '()]
        (if (empty? remaining_prime_factor_lists)
          factors
          (recur (rest remaining_prime_factor_lists) (overlap factors (first remaining_prime_factor_lists)))
        )
      )
    )
  )
)

File: ocaml/005/005.ml
let rec range i j = if i > j then [] else i :: (range (i + 1) j)

let prime_factors n =
  let rec pf c p =
    match c with
    | c when c < (p * p) -> [c]
    | c when (c mod p) = 0 -> [p] @ (pf (c / p) p)
    | _ -> pf c (p + 1)
  in
  pf n 2

let rec remove_first i l =
  match l with
  | hd::tl when hd = i -> tl
  | hd::tl -> hd::(remove_first i tl)
  | _ -> []

let overlap a b =
  let rec o a b =
    match a, b with
    | hd::tl, b -> o tl (remove_first hd b)
    | a, [] -> a
    | [], b -> b
  in
  a @ (o a b);;

Printf.printf "%d\n" (List.fold_left (fun acc x -> x * acc) 1 (List.fold_left (fun acc x -> overlap acc x) [] (List.map (fun x -> prime_factors x) (range 1 20))))
